### DISCUSSION TITLE
Problems connecting from mobile devices via studio.foxglove.dev

### DISCUSSION THREAD

**Discussion by victorjarlow** [foxglove/community#650] [Q&A]
Hi, 

I've run into the issue of foxglove studio not responding when a Rosbridge ws:// connection is opened from e.g. a mobile client. 

**Problem**
Inspecting the page loaded on the mobile client with chrome://inspect gave the following output:
![insecure_websocket](https://user-images.githubusercontent.com/97448034/223051653-8fb488b8-089b-4d1c-8ee2-a941b77db938.PNG)

Since studio.foxglove.dev uses HTTPS, plain WebSockets are not allowed by default on e.g. a mobile Android device.

**My setup:** 
* Mobile Client (Chrome 110.0.5481.65, Android 13)
* ROS2 Foxy, using rosbridge_websocket
Launched with a launch.py file with the following contents:
```python
        ...
        Node(
            package='rosbridge_server',
            executable='rosbridge_websocket',
            name='ros_bridge',
            parameters=[
                {"port": 9090},
                {"retry_startup_delay": 5.0},
                {"fragment_timeout": 600},
                {"max_message_size": 10000000},
                {"unregister_timeout": 10.0},
                {"use_compression": False}]
        ),
        Node(
            name='rosapi',
            package='rosapi',
            executable='rosapi_node'
        ),
        ...
```

**Solutions**
1. To facilitate troubleshooting, make foxglove studios be more vocal about the error and show it to the user without having to look into the console (can be a little tricky e.g. with an android client if inexperienced).
2. Allow insecure connections on the mobile client (I did not find how to do this in Android Chrome)
3. switch to Secure WebSockets (wss://) on your rosbridge_websocket server. 
    3.1 Generate a self-signed certificate e.g. using python3 (https://stackoverflow.com/a/60804101) or using OpenSSL (https://stackoverflow.com/a/41366949) (or use a signed certificate if preferred).
    3.2 Start rosbridge_websocket with certfile and keyfile parameters to enable wss:
```python
        ...
        Node(
            package='rosbridge_server',
            executable='rosbridge_websocket',
            name='ros_bridge',
            parameters=[
                {"port": 9090},
                {"retry_startup_delay": 5.0},
                {"fragment_timeout": 600},
                {"certfile": "/path/to/crt.pem},
                {"keyfile": "/path/to/key.pem},
                {"max_message_size": 10000000},
                {"unregister_timeout": 10.0},
                {"use_compression": False}]
        ),
        Node(
            name='rosapi',
            package='rosapi',
            executable='rosapi_node'
        ),
        ...
```
&nbsp;&nbsp;&nbsp;&nbsp;3.3 (If using self-signed certificate) Using the mobile client, go to https://hostname_of_rosbridge:9090 and accept the certificate.
&nbsp;&nbsp;&nbsp;&nbsp;3.4 Finally, on the mobile client, go to studio.foxglove.dev and open a Rosbridge connection, now using wss://hostname_of_rosbridge:9090, and the client should be able to connect!

**Comment by defunctzombie** [comment]
Thanks for the detailed writeup. I think you've identified the various complicated issues at play here. Ultimately this is a browser restriction for https websites being able to talk to non-"secure" resources.

One benefit of the Studio desktop app is that it does not have such annoying restrictions. We do not yet have a mobile app but this could be one approach to solving this problem.

I don't believe there is anything we can do with our web app to tell the browser to allow these types of connections from the https page.

**Reply by victorjarlow** [reply]
Thank you for the answer. A nice feature would be for the foxglove studio web app to notify the mobile user somehow then this problem occurs. Right now the page looks like it is buffering and something is about to load, I was quite confused as a first-time user of foxglove.
